src/state/tls.py

```diff
@@ -85,24 +89,33 @@ def from_charm(
         return cls(
             hostnames=hostnames,
             tls_cert_and_ca_chain=tls_cert_and_ca_chain,
-            private_key=private_key,
+            private_key=str(private_key),
         )
 
     # Validation is done in this method instead of using a pydantic model because
     # there are cases where we need to validate the state but we don't need the state instance.
     @classmethod
-    def validate(cls, charm: ops.CharmBase, certificates: TLSCertificatesRequiresV4) -> None:
+    def validate(
+        cls,
+        charm: ops.CharmBase,
+        certificates: TLSCertificatesRequiresV4,
+        allow_no_certificates: bool = False,
+    ) -> None:
         """Validate the precondition to initialize this state component.
 
         Args:
             charm: The haproxy charm.
             certificates: TLS certificates requirer class.
+            allow_no_certificates: If the charm can proceed without requesting any certificates.
 
         Raises:
             TLSNotReadyError: if the charm is not ready to handle TLS.
         """
         tls_requirer_integration = charm.model.get_relation(certificates.relationship_name)
-        if not certificates.certificate_requests:
+        if allow_no_certificates:
```

Comment:
Can you add unit tests covering this piece of code?

Reply:
added!